client/app/services/parameters/Parameter.js

```diff
@@ -0,0 +1,160 @@
+import { isNull, isObject, isFunction, isUndefined, isEqual, has, omit, isArray, each } from 'lodash';
+import {
+  TextParameter, NumberParameter, EnumParameter, QueryBasedDropdownParameter,
+  DateParameter, DateRangeParameter,
+} from '.';
+
+class Parameter {
+  constructor(parameter, parentQueryId) {
+    this.title = parameter.title;
+    this.name = parameter.name;
+    this.type = parameter.type;
+    this.global = parameter.global; // backward compatibility in Widget service
+    this.parentQueryId = parentQueryId;
+
+    // Used for meta-parameters (i.e. dashboard-level params)
+    this.locals = [];
+
+    // Used for URL serialization
+    this.urlPrefix = 'p_';
+  }
+
+  static create(param, parentQueryId) {
+    switch (param.type) {
+      case 'number':
+        return new NumberParameter(param, parentQueryId);
+      case 'enum':
+        return new EnumParameter(param, parentQueryId);
+      case 'query':
+        return new QueryBasedDropdownParameter(param, parentQueryId);
+      case 'date':
+      case 'datetime-local':
+      case 'datetime-with-seconds':
+        return new DateParameter(param, parentQueryId);
+      case 'date-range':
+      case 'datetime-range':
+      case 'datetime-range-with-seconds':
+        return new DateRangeParameter(param, parentQueryId);
+      default:
+        return new TextParameter({ ...param, type: 'text' }, parentQueryId);
+    }
+  }
+
+  static getExecutionValue(param, extra = {}) {
+    if (!isObject(param) || !isFunction(param.getExecutionValue)) {
+      return null;
+    }
+
+    return param.getExecutionValue(extra);
+  }
+
+  static setValue(param, value) {
+    if (!isObject(param) || !isFunction(param.setValue)) {
+      return null;
+    }
+
+    return param.setValue(value);
+  }
+
+  get isEmpty() {
+    return isNull(this.getExecutionValue());
+  }
+
+  get hasPendingValue() {
+    return this.pendingValue !== undefined && !isEqual(this.pendingValue, this.normalizedValue);
+  }
+
+  /** Get normalized value to be used in inputs */
+  get normalizedValue() {
+    return this.$$value;
+  }
+
+  // TODO: Remove this property when finally moved to React
+  get ngModel() {
+    return this.normalizedValue;
+  }
```

Comment:
Do we use this property anywhere? I mean - can it be removed right now?

Reply:
I thought it could be something used by Angular internally :sweat_smile:, but from a quick search on when it was introduced it does seem we don't use it anymore, so yeah, we can get rid of that :scissors: 

Thanks!